### DISCUSSION TITLE
npm install is slow ?

### DISCUSSION THREAD

**Discussion by cburkins** [dokku/dokku#5371] [General]
I'm not sure where to start here.... so looking to build my experience so I can ask a more intelligent question

On my local machine, the "npm install" for my node_modules only takes about 15s.

When I push my branch up to my dokku installation, the same npm install takes ~4m (so about =16x longer).   Not a big deal obviously, as 4m is only 4m but I'm curious to know how I can improve this.

* Does npm cache better locally ?
* Does npm cache at all in my dokku installation ?  The logs indicate that it does.

**Comment by cburkins** [comment]
To be fair, this isn't a really apples-to-apples comparison:
my local machine: M1 MacBookPro
my remote dokku machine: AWS t2.medium EC2

So figured I would manually load the repo into my home directory on that remote dokku machine and time it:
"npm install" on the following machines:
M1 MacBookPro: 0m09s (clear my npm cache by removing ~/.npm first)
git push to dokku: ~3minutes (doesn't give me exact time)
running manual npm install on *same* dokku machine: 0m15s (see NOTE below)

Feels like dokku doesn't really cache npm packages ?

NOTE: I can get the manual "npm install" to run much longer on the remote dokku machine if I use "--prefer-online" option.  Much more in-line with the dokku "npm install".   Which reinforces that dokku is not cacheing ?

**Reply by josegonzalez** [reply]
The default nodejs buildpack should cache dependencies, and it shouldn't be much slower than a deploy to heroku (depending on your instance resources of course).

I highly doubt an install without any cache takes 9 seconds (did you delete your node_modules directory?) if the entire deploy takes 3 minutes.

One thing to note is that a deploy does more than just build your app, and may be slower as it starts/stops ephemeral build containers and then starts your actual running containers. Without a recording of the deploy (you can get that by using something like asciinema), I can't really diagnose why it might be slower than expected.

**Comment by cburkins** [comment]
Figured it out....    I've got a node/express app, with a React client in /client.    I build /client with "cd client && npm install".  To get that to cache, I need to include the following in the root /package.json

```
    "cacheDirectories": [
        "client/node_modules",
        "./node_modules"
    ]
```

That's the good news.   

The bad news is that's it doesn't really go any faster.  It takes nearly as long to cache the node_modules than it does to build it.   Recently switched from CRA to Vite, and Vite is a good bit smaller to install.

Oh well. 

I've got my builds down to about ~4minutes
